src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                var analyzers = stateSets.Select(s => s.Analyzer).ToImmutableArrayOrEmpty();
+                return CreateAnalyzerDriverAsync(project, analyzers, includeSuppressedDiagnostics, cancellationToken);
+            }
+
+            public async Task<CompilationWithAnalyzers> CreateAnalyzerDriverAsync(
+                Project project, ImmutableArray<DiagnosticAnalyzer> analyzers, bool includeSuppressedDiagnostics, CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
+                {
+                    return null;
+                }
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+                // Create driver that holds onto compilation and associated analyzers
+                return CreateAnalyzerDriver(
+                    project, compilation, analyzers, logAnalyzerExecutionTime: false, reportSuppressedDiagnostics: includeSuppressedDiagnostics);
+            }
+
+            private CompilationWithAnalyzers CreateAnalyzerDriver(
+                Project project,
+                Compilation compilation,
+                ImmutableArray<DiagnosticAnalyzer> analyzers,
+                bool logAnalyzerExecutionTime,
+                bool reportSuppressedDiagnostics)
+            {
+                // PERF: there is no analyzers for this compilation.
+                //       compilationWithAnalyzer will throw if it is created with no analyzers which is perf optimization.
+                if (analyzers.IsEmpty)
+                {
+                    return null;
+                }
+
+                Contract.ThrowIfFalse(project.SupportsCompilation);
+                AssertCompilation(project, compilation);
+
+                var analysisOptions = GetAnalyzerOptions(project, logAnalyzerExecutionTime, reportSuppressedDiagnostics);
+
+                // Create driver that holds onto compilation and associated analyzers
+                return compilation.WithAnalyzers(analyzers, analysisOptions);
+            }
+
+            private CompilationWithAnalyzersOptions GetAnalyzerOptions(
+                Project project,
+                bool logAnalyzerExecutionTime,
+                bool reportSuppressedDiagnostics)
+            {
+                // in IDE, we always set concurrentAnalysis == false otherwise, we can get into thread starvation due to
+                // async being used with syncronous blocking concurrency.
+                return new CompilationWithAnalyzersOptions(
+                    options: new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Workspace),
+                    onAnalyzerException: GetOnAnalyzerException(project.Id),
+                    analyzerExceptionFilter: GetAnalyzerExceptionFilter(project),
+                    concurrentAnalysis: false,
```

Comment:
> until CompilationWithAnalyzer uses async based concurrency rather than synchronous lock based concurrency, it is too dangerous to be used in IDE

We already do so in v1 engine for FixAll code path - and that drastically improved the FixAll perf in Update2 compared to Update1. Note that this is being done from the UI thread. If we pass ConcurrentAnalysis = false for FixAll code path, it will go back to Update1 perf, which was basically unusable. Can you try doing a FixAll in decent sized solution with v2 engine and see perf (or run the VS integration test CSharpFixAllOccurrences.xml from [this PR](https://github.com/dotnet/roslyn-internal/pull/805) locally)?

> until CompilationWithAnalyzer uses async based concurrency rather than synchronous lock based concurrency

I explicitly tried replacing all locks with SemaphoreSlim DisposableWaitAsync in CompilationWithAnalyzers but that caused large VM regressions - see [this PR](https://github.com/dotnet/roslyn/pull/6903)